### DISCUSSION TITLE
Is an Assembly function to quickly compute XSize*y+x available in C or can it be imported somehow from CC65 source code?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#2519] [Q&A]
Hi everyone! I am a lousy Assembly coder and I need to have a quick function that computes XSize*y+x, for any given x,y.
Such function should already be available in the source code of CC65 to do things such as `gotoxy`.
So far I have not been able to find it. I can try harder and find it, myself but that is not the only question here. I wonder whether we can import and reuse this (and possibly) other Assembly snippets of Assembly code in CC65 source code in C programs and how (other than copy pasting them)?

**Comment by mrdudz** [comment]
> Hi everyone! I am a lousy Assembly coder and I need to have a quick function that computes XSize*y+x, for any given x,y.
Such function should already be available in the source code of CC65 to do things such as gotoxy.

In case of the CBM targets, this function is in the kernal ROM (the PLOT function). (And it does not actually multiply, it uses a table based approach)

> I wonder whether we can import and reuse this (and possibly) other Assembly snippets of Assembly code in CC65 source code in C programs and how (other than copy pasting them)?

I don't understand the question - what exactly do you want to do?

**Reply by Fabrizio-Caruso** [reply]
I would like to know how to quickly compute XSize*Y+X (not standardScreenArea+XSize*Y+X because I use a double buffer) on several 6502-targets. I need to use such functions from C. I am a poor Assembly coder. Would you be so nice to tell me how I can do it at least on the CBM targets? A more general solution would be necessary for other targets but it is OK to start with CBM.

P.S.: The goal is to integrate it in the C APIs of my Cross-Lib framework.

**Comment by mrdudz** [comment]
just use a lookup table

; X in A, Y in X
clc
adc table,x

with a table like
.repeat n,maxlines
.byte (40 * n)
.endrepeat

**Comment by Fabrizio-Caruso** [comment]
Thanks @mrdudz 

Is `.byte (40*n)` going to generate a 16-bit word ? 40*n>255 
Should it rather be` .word` ??

I am not sure how I can use this from C. How do I return tha values? Should I use embedded Assembly with some __asm__ or a separate .s file?

Something like
```
uint16_t screen_offset(uint8_t x, uint8_t y)
{
    asm_screen_offset(x,y);
    return ??   
}
```
or
uint16_t screen_offset(uint8_t x, uint8_t y)
```
{
__asm__('...');
}
```

**Reply by rofl0r** [reply]
https://codebase64.org/doku.php?id=base:6502_6510_maths

**Reply by mrdudz** [reply]
> https://codebase64.org/doku.php?id=base:6502_6510_maths

Uhm, no. You don't want a full blown multiplication for screen output, ever :)

**Comment by mrdudz** [comment]
> Is .byte (40*n) going to generate a 16-bit word ? 40*n>255 Should it rather be .word ??

of course. As always, look at an existing implementation of a function with the same signature... (eg gotoxy). Anyway, something like this:

```
; uint16_t screen_offset(uint8_t x, uint8_t y)
.export screen_offset

screen_offset:
    jsr     popa            ; Get Y
    sta    tmp1
    jsr     popa            ; Get X
    ldx    tmp1
    clc
    adc tablelo,x
    pha
    lda #0
    adc tablehi,x
    tax ; highbyte to x
    pla ; lowbyte in A
    rts
tablelo:
.repeat n,maxlines
.byte < (40 * n)
.endrepeat
tablehi:
.repeat n,maxlines
.byte > (40 * n)
.endrepeat
```

totally untested of course :) it can probably be optimized as well (using fastcall etc).

Depending on how fast you need the result, you can add the highbyte of the screen start address to the table values (and use two tables for doublebuffer) or just add it after the lookup.

**Comment by groessler** [comment]
;
;multiplikationsroutine entnommen  mc 2/83
;von k.-g. meyer
;
mul8:   lda     #0
         ldy     #8
loop8:  asl     a
         rol     mulor
         bcc     lab1
         clc
         adc     mulnd
         bcc     lab1
         inc     mulor
lab1:   dey
         bne     loop8
         rts
;

**Comment by Fabrizio-Caruso** [comment]
Thanks everyone! @mrdudz looks the fastest solution because it exploits the assumption that 40 is a factor.
